### PR TITLE
add more tests

### PR DESCRIPTION
**Pull request by Akuli**
extracted from #41

### CODE REVIEW

**Comment by Akuli** [comment]
I think I might move go `tests/go` somewhere else:
- Pytest is slow because it looks at every subdirectory of `tests/`
- `black tests/` also reformats a Python file that Go comes with

File: README.md [hunk 1]
```diff
@@ -2,3 +2,31 @@
...
+TODO: add instructions for looking at coverage files on Windows
+
+To run tests, first download the langservers you need.
+You can mostly read `.github/workflows/test.yml`, but the Go langserver is a bit of a gotcha.
+You will need to install go from https://golang.org/,
+because the one from `sudo apt install golang` is too old.
+Extract it inside where you cloned `sansio-lsp-client`
+so that you get an executable named `sansio-lsp-client/go/bin/go`.
+
+    $ tar xf /blah/blah/Downloads/go1.16.5.linux-amd64.tar.gz
```
Comment (Akuli) [inline README.md:28]:
```suggestion
    $ tar xf ~/Downloads/go1.16.5.linux-amd64.tar.gz
```

File: tests/test_actual_langservers.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+
+# TODO: More stuff coming from halfbrained pull requests
+METHOD_COMPLETION = "completion"
+RESPONSE_TYPES = {METHOD_COMPLETION: lsp.Completion}
+
+
+def get_meth_text_pos(text, method):
+    """ searches for line: `<code> #<method>-<shift>`
+          - example: `sys.getdefaultencoding()
```
Comment (Akuli) [inline tests/test_actual_langservers.py:26]: This docstring has suffered from some copy/pasta spaghetti

File: tests/test_actual_langservers.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+    def exit_cleanly(self):
+        #        if self.msgs:
+        #            # Not necessarily error, gopls sends logging messages for example
+        #            print(
+        #                "* unprocessed messages: " + pprint.pformat(self.msgs)
+        #            )
+
+        assert self.lsp_client.state != lsp.ClientState.NOT_INITIALIZED
+        assert self.lsp_client.state != lsp.ClientState.WAITING_FOR_INITIALIZED
```
Comment (Akuli) [inline tests/test_actual_langservers.py:148]: Maybe this should assert specifically what state it's in

File: tests/test_actual_langservers.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+            raise ValueError(langserver_name)
+
+
+def _needs_clangd(version):
+    return pytest.mark.skipif(
+        not list(langserver_dir.glob(f"clangd_{version}.*")),
+        reason=f"clangd {version} not found",
+    )
+
+
```
Comment (Akuli) [inline tests/test_actual_langservers.py:368]: Not used

```suggestion
```

**Comment by Akuli** [comment]
@PurpleMyst Can you review these PRs this weekend?

**Comment by PurpleMyst** [comment]
Yep, I'm reviewing them as we speak. Do I need to review the draft PR also or?

File: tests/test_actual_langservers.py [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+            for ev in events:
+                self.msgs.append(ev)
+                self._try_default_reply(ev)
+
+    def _try_default_reply(self, msg):
+        if isinstance(msg, lsp.ShowMessageRequest):
+            msg.reply()
+
+    def wait_for_message_of_type(self, type_, timeout=5):
+        end_time = time.time() + timeout
```
Comment (PurpleMyst) [inline tests/test_actual_langservers.py:113]: Maybe a place for [time.monotonic](https://devdocs.io/python~3.9/library/time#time.monotonic)?

**Review by PurpleMyst — APPROVED** [review]
LGTM, if a bit hardcode-y. Maybe this is fine for tests, but perhaps a small config file (even in python to avoid having a weird pseudo-programming-language) that handles which LSPs to test and how could be interesting for the future.

**Comment by Akuli** [comment]
The draft PR is not ready for review. I'd rather get other PRs merged first, so that the draft PR will become easier to review.

**Comment by PurpleMyst** [comment]
Alright, let me know when the draft PR is ready for review and I'll review it. Awesome work all around, by the way!

Side-note: I joined the IRC channel on freenode today and I saw that there was no one there anymore. Have you guys moved to another channel, like on libera.chat, or maybe to another platform?

**Comment by Akuli** [comment]
Yeah, same channel names but on libera.

**Pull request merged**